### PR TITLE
[Feature] Support DSDL Dataset

### PR DESCRIPTION
**Pull request by wufan-tb**
- support dsdl seg dataset 
- add dsdl dataset citest
- validated accuracy on voc2012 and cityscapes

### CODE REVIEW

File: configs/dsdl/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  ```
+  git clone https://github.com/opendatalab/dsdl-sdk.git -b schema-dsdl
+  cd dsdl-sdk
+  python setup.py install
+  ```
+
+- install mmdet and pytorch:
+  please refer this [installation documents](https://mmdetection.readthedocs.io/en/3.x/get_started.html).
+
```
Comment (xiexinch) [inline configs/dsdl/README.md:35]: Is it necessary to install mmdet?

File: configs/dsdl/cityscapes.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    '../_base_/models/deeplabv3_r50-d8.py', '../_base_/default_runtime.py',
+    '../_base_/schedules/schedule_40k.py'
+]
+
+crop_size = (512, 1024)
+data_preprocessor = dict(size=crop_size)
+model = dict(data_preprocessor=data_preprocessor)
+# dataset settings
+dataset_type = 'DSDLSegDataset'
+data_root = 'data/CityScapes'
```
Comment (xiexinch) [inline configs/dsdl/cityscapes.py:11]:
```suggestion
data_root = 'data/cityscapes'
```

File: configs/dsdl/cityscapes.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+img_prefix = 'original'
+train_ann = 'dsdl/set-train/train.yaml'
+val_ann = 'dsdl/set-val/val.yaml'
```
Comment (xiexinch) [inline configs/dsdl/cityscapes.py:14]: Should we create the `dsdl` folder at mmsegmentation root path or at the datasets root path?

File: configs/dsdl/cityscapes.py (hunk 2 above)
Comment (xiexinch) [reply configs/dsdl/cityscapes.py:11]: I think this path should correspond to the one in the documentation.  https://mmsegmentation.readthedocs.io/en/latest/user_guides/2_dataset_prepare.html

File: configs/dsdl/voc.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+_base_ = [
+    '../_base_/models/deeplabv3_r50-d8.py', '../_base_/default_runtime.py',
+    '../_base_/schedules/schedule_20k.py'
+]
+
+# dataset settings
+dataset_type = 'DSDLSegDataset'
+data_root = 'data/VOC12-seg'
+img_prefix = 'original'
```
Comment (xiexinch) [inline configs/dsdl/voc.py:9]: Is 'original' a keyword needed inside DSDL?

File: mmseg/datasets/dsdl.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                 used_labels: Optional[Sequence] = None,
+                 **kwargs) -> None:
+
+        if DSDLDataset is None:
+            raise RuntimeError(
+                'Package dsdl is not installed. Please run "pip install dsdl".'
+            )
+        self.used_labels = used_labels
+
+        loc_config = dict(type='LocalFileReader', working_dir='')
```
Comment (xiexinch) [inline mmseg/datasets/dsdl.py:38]: Could it be customized by config file?

File: mmseg/datasets/dsdl.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        if kwargs.get('data_root'):
+            kwargs['ann_file'] = os.path.join(kwargs['data_root'],
+                                              kwargs['ann_file'])
+        self.required_fields = ['Image', 'LabelMap']
```
Comment (xiexinch) [inline mmseg/datasets/dsdl.py:42]: This variable doesn't seem to be used by other functions, do we need to make it a property of the class?

File: configs/dsdl/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
+# DSDL: Standard Description Language for DataSet
+
+<!-- [ALGORITHM] -->
```
[the remark this answers is not in the document]
Comment (wufan-tb) [reply configs/dsdl/README.md:3]: fixed

File: mmseg/datasets/dsdl.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            location_config=loc_config,
+            required_fields=self.required_fields,
+            specific_key_path=specific_key_path,
+            transform=pre_transform,
+        )
+        BaseSegDataset.__init__(self, **kwargs)
+
+    def load_data_list(self) -> List[Dict]:
+        """Load data info from an dsdl yaml file named as ``self.ann_file``
+
```
[the remark this answers is not in the document]
Comment (wufan-tb) [reply mmseg/datasets/dsdl.py]: fixed

File: configs/dsdl/README.md (hunk 1 above)
Comment (wufan-tb) [reply configs/dsdl/README.md]: fixed

File: configs/dsdl/cityscapes.py (hunk 2 above)
Comment (wufan-tb) [reply configs/dsdl/cityscapes.py]: fixed

File: configs/dsdl/cityscapes.py (hunk 3 above)
Comment (wufan-tb) [reply configs/dsdl/cityscapes.py:14]: this path will concate with root_path, so the actually path is "data/cityscapes/dsdl/set-train/train.yaml".

furthermore,  here is a typical example of dsdl dataset's file struct :
```
data
└── VOC12-seg
    ├── dsdl
    └── original
```

File: configs/dsdl/voc.py (hunk 4 above)
Comment (wufan-tb) [reply configs/dsdl/voc.py:9]: no, users should change it by actual path. (or no img_prefix, decied by it's file sturct)

File: mmseg/datasets/dsdl.py (hunk 5 above)
Comment (wufan-tb) [reply mmseg/datasets/dsdl.py:38]: there is no need to customize this parameter.
we use 'LoadImageFromFile' in mmengine (not this 'LocalFileReader' in dsdl) to load image files, so only send an empty loc_config to initialize DSDLDataset class.

File: mmseg/datasets/dsdl.py (hunk 6 above)
Comment (wufan-tb) [reply mmseg/datasets/dsdl.py]: fixed

File: mmseg/datasets/dsdl.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            be loaded by it's field name.
+        pre_transform(dict): pre-transform functions before loading.
+    """
+
+    METAINFO = {}
+
+    def __init__(self,
+                 specific_key_path: Dict = {},
+                 pre_transform: Dict = {},
+                 used_labels: Optional[Sequence] = None,
```
Comment (xiexinch) [inline mmseg/datasets/dsdl.py:29]: Might introduce `used_labels` at docstring.

**Review by xiexinch — APPROVED** [review]

**Pull request merged**